### PR TITLE
camera intrinsics calibration

### LINKED ISSUE #105: Calibration in ROS
Hi all!

I have been trying to calibrate the Azure Kinect following the tutorial specified at:

[http://wiki.ros.org/camera_calibration/Tutorials/MonocularCalibration](http://wiki.ros.org/camera_calibration/Tutorials/MonocularCalibration)

From what I understand, what it does is that it gets calibration information and then communicates with ROS service /camera/set_camera_info to change the intrinsic parameters of the camera. 

I didn't find the camera/set_camera_info service being published by the camera node and hence when I click save after calibrating (following the tutorial), I get the following error:

```
Traceback (most recent call last):
  File "/opt/ros/melodic/lib/python2.7/dist-packages/camera_calibration/camera_calibrator.py", line 252, in on_mouse
    if self.do_upload():
  File "/opt/ros/melodic/lib/python2.7/dist-packages/camera_calibration/camera_calibrator.py", line 209, in do_upload
    response = self.set_camera_info_service(info)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 439, in __call__
    return self.call(*args, **kwds)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 499, in call
    service_uri = self._get_service_uri(request)
  File "/opt/ros/melodic/lib/python2.7/dist-packages/rospy/impl/tcpros_service.py", line 467, in _get_service_uri
    raise ServiceException("service [%s] unavailable"%self.resolved_name)
rospy.service.ServiceException: service [/camera/set_camera_info] unavailable

```

Is there a way out of it? Kindly suggest any other way that has worked for you. 

ROS Version: Melodic
Ubuntu 18.04

Ros Services running:

> /azure_kinect_ros_driver/get_loggers
/azure_kinect_ros_driver/set_logger_level
/depth/image_raw/compressed/set_parameters
/depth/image_raw/compressedDepth/set_parameters
/depth/image_raw/theora/set_parameters
/depth_to_rgb/image_raw/compressed/set_parameters
/depth_to_rgb/image_raw/compressedDepth/set_parameters
/depth_to_rgb/image_raw/theora/set_parameters
/ir/image_raw/compressed/set_parameters
/ir/image_raw/compressedDepth/set_parameters
/ir/image_raw/theora/set_parameters
/joint_state_publisher/get_loggers
/joint_state_publisher/set_logger_level
/rgb/image_raw/compressed/set_parameters
/rgb/image_raw/compressedDepth/set_parameters
/rgb/image_raw/theora/set_parameters
/rgb_to_depth/image_raw/compressed/set_parameters
/rgb_to_depth/image_raw/compressedDepth/set_parameters
/rgb_to_depth/image_raw/theora/set_parameters
/robot_state_publisher/get_loggers
/robot_state_publisher/set_logger_level
/rosout/get_loggers
/rosout/set_logger_level
/rostopic_20269_1575590186848/get_loggers
/rostopic_20269_1575590186848/set_logger_level

Ros Nodes running:

> /azure_kinect_ros_driver
/joint_state_publisher
/robot_state_publisher
/rosout
/rostopic_20269_1575590186848
 
Appreciate your time and help. Thanks!

### LINKED ISSUE #199: driver publishes unrectified images by default
**Describe the bug**
~~The intrinsic calibration that is used to project colour and depth pixels to a 3D point cloud is wrong or not properly processed.~~ According to https://github.com/microsoft/Azure-Kinect-Sensor-SDK/issues/726 and https://github.com/microsoft/Azure_Kinect_ROS_Driver/issues/105 it is neither possible to update the intrinsic parameters on the camera, nor does the ROS node support calibration.

Edit: The `driver.launch` indeed publishes the correct intrinsics together with the raw unrectified images. These images have to be rectified manually. However, the driver does not provide currently a way to update these intrinsics.

**To Reproduce**
1. set up a scene with AprilTags
2. visualise the point cloud as well as the estimated tag frames in RViz
3. see that frames do not align with depth

**Expected behavior**
The intrinsic calibration should provide proper 3D point clouds.

**Screenshots**

The following images show a point cloud together with the coordinate frames of two AprilTags in RViz. One AprilTag is flat on the table and another one is attached to the side of box and thus orthogonal to the table.

With the Kinect factory calibration, those points and coordinate frame do not align:
atag_kinect1
atag_kinect2

As a reference, here is the data from an Asus Xtion Pro Live using the `openni2_camera` package and [manual intrinsic calibration](http://wiki.ros.org/camera_calibration/Tutorials/MonocularCalibration). As you can see, these frames perfectly align with the point cloud.
atag_openni1
atag_openni2

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Version 20.04

### PR DESCRIPTION
**Pull request by christian-rauch**
### Description of the changes:
- add the camera_info_manager to store/load camera parameters
- add rgbd_launch for image rectification
- ~~all of the changes in https://github.com/microsoft/Azure_Kinect_ROS_Driver/pull/180~~

### Required before submitting a Pull Request:
- [x] I reviewed [CONTRIBUTING.md](https://github.com/microsoft/Azure_Kinect_ROS_Driver/blob/master/CONTRIBUTING.md)
- [x] I [built my changes](https://github.com/microsoft/Azure_Kinect_ROS_Driver/blob/master/docs/building.md) locally
- [x] I tested my changes with a device
- [ ] I changed both the ROS1 and ROS2 branches

### I tested changes on: 
- [ ] Windows
- [x] Linux
- [x] ROS1
- [ ] ROS2

With the standard monocular calibration described in http://wiki.ros.org/camera_calibration/Tutorials/MonocularCalibration can now achieve a much better AprilTag to point cloud alignment:

atag_kinect_calib1
atag_kinect_calib2

You have to use the new `kinect_rgbd.launch` launch file to rectify the images.

Note:
To calibrate the IR sensor, you likely have to reduce the scaling to something around  0.2
```
roslaunch azure_kinect_ros_driver kinect_rgbd.launch ir_mono8_scaling_factor:=0.2
```
to prevent saturation of the IR image.

### CODE REVIEW

File: include/azure_kinect_ros_driver/k4a_ros_device_params.h [hunk 1]
```diff
@@ -33,6 +33,7 @@
 #define ROS_PARAM_LIST                                                                                                 \
   LIST_ENTRY(sensor_sn, "The serial number of the sensor this node should connect to.", std::string, std::string(""))  \
   LIST_ENTRY(depth_enabled, "True if depth camera should be enabled", bool, true)                                      \
+  LIST_ENTRY(depth_unit, "Depth distance units. Options are: metre, millimetre", std::string, std::string("metre"))    \
```
Comment (ooeygui) [inline include/azure_kinect_ros_driver/k4a_ros_device_params.h]: Can you add support for both spellings?

**Review by ooeygui — DISMISSED** [review]

File: include/azure_kinect_ros_driver/k4a_ros_device_params.h (hunk 1 above)
Comment (christian-rauch) [reply include/azure_kinect_ros_driver/k4a_ros_device_params.h]: I changed that to `32FC1` and `16UC1` now. These are the unit conventions used by ROS and this should make it more clear in which format the depth will be provided. I changed that also in the original PR https://github.com/microsoft/Azure_Kinect_ROS_Driver/pull/180. Not sure if you want to merge that first, or both at the same time with this one.

**Comment by christian-rauch** [comment]
Some more information about the different intrinsic parameters. I am using the `kinect_rgbd.launch` launch file which uses `rgbd_launch` for rectification and registration. The raw images are published to `/rgb/image_raw` and `/ir/image_raw` (as before) and the rectified images are published to `/rgb/image_rect_color` and `/ir/image_rect_ir`. 

The ROS [`image_proc` nodelets](image_proc) can use the factory calibration for rectification and registration, but the **factory** and **manual** calibration differs in how the RGB and IR images are warped. The factory calibration fully warps the image such that all lines are straight but adds an irregular black border to the image due to the full warping. The manual calibration only partially warps the image such that the borders are kept. Hence, the manual calibration is useful if you intend to post-process the rectified image (e.g. edge and keypoint detection) and do not want to deal with manual filtering of the output. Of course, manual calibration will also be required if you intend to change the optics of the camera.

The following raw image (`/rgb/image_raw`):
image2_raw

will be rectified with the factory calibration (`/rgb/image_rect_color`) as:
image2_rect_factory

and the [manual calibration](http://wiki.ros.org/camera_calibration/Tutorials/MonocularCalibration) will generate:
image2_rect_manual

**Review by amelhassan — DISMISSED** [review]

**Comment by amelhassan** [comment]
I also verified using Melodic on Windows that this functions, thanks for your contribution @christian-rauch. Do you know if its possible to have something similar for ROS2 (it doesn't look like the rgdb_launch package has a ROS2 equivalent)? Also it would be helpful if you included some of your quick comments on using the kinect_rgbd.launch in the using.md (when to use it, and mention needing the rgdb_launch package).

**Comment by christian-rauch** [comment]
I added documentation for the launch files and calibration to `usage.md`.

A quick look shows that there is at least an active ros2 branch for the [`camera_info_manager`](https://github.com/ros-perception/image_common/tree/ros2/camera_info_manager) and for the [`image_proc`](https://github.com/ros-perception/image_pipeline/tree/ros2/image_proc) that does the rectification. That should cover the basic calibration and rectification features.
I haven't looked into implementing this for the ROS2 as our setup does not use ROS2 yet. Maybe someone at Microsoft can adapt this to the ROS2 version of the driver?

**Review by amelhassan — APPROVED** [review]

**Comment by amelhassan** [comment]
Great, thank you for those changes. Our team will look into creating a similar change for the ROS2 version. I created issue #211 to represent that.

**Comment by bmegli** [comment]
> The ROS [image_proc nodelets](https://github.com/microsoft/Azure_Kinect_ROS_Driver/pull/image_proc) can use the factory calibration for rectification and registration, but the factory and manual calibration differs in how the RGB and IR images are warped. The factory calibration fully warps the image such that all lines are straight but adds an irregular black border to the image due to the full warping. The manual calibration only partially warps the image such that the borders are kept. Hence, the manual calibration is useful if you intend to post-process the rectified image (e.g. edge and keypoint detection) and do not want to deal with manual filtering of the output. Of course, manual calibration will also be required if you intend to change the optics of the camera.

For the record:
- it is possible to implement factory calibration behavior like manual calibration
- by using  OpenCV's `getOptimalNewCameraMatrix` with `alpha == 0`
  - to determine CameraInfo.P matrix
  - when publishing CameraInfo
- example here
  - https://github.com/microsoft/Azure_Kinect_ROS_Driver/issues/212#issuecomment-1181486020

It should be relatively easy to:
- have both behaviors depending on roslaunch param
  - e.g. `arg name="useOptimalRectifcationCameraMatrix" default="false" />`
  - publish as is now (copy K to P) when `false`
  - determine new P intrinsics using OpenCV when `true`

**Pull request merged**